### PR TITLE
Changes to BlobStoreCompactor to handle TTL update records

### PR DESCRIPTION
**Pull request by vgkholla**
Tests already cover a lot of the new logic. Will add some more tests for corner cases with TTL updates (current test already cover non corner case TTL updates)

### CODE REVIEW

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 1]
```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+   * @throws StoreException if there are problems reading the index
+   */
+  private boolean isTtlUpdateEntryValid(StoreKey key, Offset indexSegmentStartOffset) throws StoreException {
+    boolean valid = false;
+    //  A TTL update entry is "valid" if the corresponding PUT is still alive
+    // The PUT entry, if it exists, must be "before" this TTL update entry.
+    FileSpan srcSearchSpan = new FileSpan(srcIndex.getStartOffset(), indexSegmentStartOffset);
+    IndexValue srcValue = srcIndex.findKey(key, srcSearchSpan, EnumSet.of(PersistentIndex.IndexEntryType.PUT));
+    if (srcValue == null) {
+      // PUT is not in the source - therefore can't be in target. This TTL update can be cleaned up
```
Comment (pnarayanan) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: when you say "therefore cannot be in target", do you mean "therefore cannot go into the target"? (just a little confusing).

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 2]
```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+      logger.trace("TTL update of {} in segment with start offset {} in {} is not valid the corresponding PUT entry "
+          + "does not exist anymore", key, indexSegmentStartOffset, storeId);
+    } else {
+      IndexValue tgtValue = tgtIndex.findKey(key, null, EnumSet.of(PersistentIndex.IndexEntryType.PUT));
+      if (tgtValue == null && isOffsetInCurrentCycle(srcValue.getOffset())) {
+        // exists in src but not in tgt. This can happen either because
+        // 1. The FileSpan to which srcValue belongs is not under compaction
+        // 2. srcValue will be compacted in this cycle
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Hard to understand 2, more comments for 2 would be great.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 3]
```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+  }
+
+  /**
+   * @param key the {@link StoreKey} to check
+   * @param searchStartOffset the start offset of the search for delete entry
+   * @param searchEndOffset the end offset of the search for delete entry
+   * @return {@code true} if the key has been deleted
+   * @throws StoreException if there are any problems using the index
+   */
+  private boolean isDeleted(StoreKey key, Offset searchStartOffset, Offset searchEndOffset) throws StoreException {
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Add more comments why a put can be deleted, mention the fact that delete may not in current compaction range. 

Rename it to `hasDeleteEntryNotUnderCompaction` ?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 4]
```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+      }
+    }
+    return valid;
+  }
+
+  /**
+   * @param offset the {@link Offset} to check
+   * @return {@code true} if the offset is being compacted in the current cycle.
+   */
+  private boolean isOffsetInCurrentCycle(Offset offset) {
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: rename to `isOffsetUnderCompaction`?

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 5]
```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+
+  /**
+   * Determines if {@code copyCandidate} is a duplicate.
+   * @param copyCandidate the {@link IndexEntry} to check
+   * @param duplicateSearchSpan the search span for duplicates in the source index
+   * @param indexSegmentStartOffset the start offset of the {@link IndexSegment} being processed
+   * @param checkAlreadyCopied if {@code true}, checks if {@code copyCandidate} has already been copied.
+   * @return {@code true} if {@code copyCandidate} is a duplicate of something that exists or has been copied
+   */
+  private boolean isDuplicate(IndexEntry copyCandidate, FileSpan duplicateSearchSpan, Offset indexSegmentStartOffset,
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Add high level comments why duplicate exist.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 2 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: I added some more comments. Let me know if more info can be provided

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 3 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Made the changes discussed offline

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 4 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Done

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 5 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Done

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 6]
```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+          NavigableSet<IndexValue> values = indexSegment.find(indexEntry.getKey());
+          IndexValue secondVal = values.lower(values.last());
+          if (secondVal != null && secondVal.isFlagSet(IndexValue.Flags.Ttl_Update_Index) && isTtlUpdateEntryValid(
+              indexEntry.getKey(), indexSegment.getStartOffset())) {
+            validEntries.add(new IndexEntry(indexEntry.getKey(), secondVal));
           }
+        } else {
+          // if this delete cannot be counted and there is a corresponding unexpired PUT/TTL update entry in the same
+          // index segment, we will need to add it.
+          addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: I didn't want to create the extra list required to do so but let's discuss offline (this is similar to what `IndexSegment#getIndexEntriesSince()` does)

**Comment by vgkholla** [comment]
@zzmao comments addressed. Just have to discuss one comment

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 7]
```diff
@@ -590,140 +590,286 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
+            } else {
+              logger.trace("{} in index segment with start offset {} in {} is not valid because it is a deleted PUT",
+                  indexEntry, indexSegment.getStartOffset(), storeId);
+            }
           } else {
-            logger.trace("Skipping {} in index segment with start offset {} in {} because it is a deleted PUT",
-                indexEntry, indexSegmentStartOffset, storeId);
+            // valid PUT entry
+            // Add all values in this index segment (to account for the presence of TTL updates)
+            addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: If `oneEntryPerKey`, `PUT` should be the latest and only one in the `indexSegment`. How `TTL` presents in this `indexSegment` ?

Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: If you look at the code in `IndexSegment#eliminateDuplicates()`, these are the possible input and output combinations
1. PUT entry only -> PUT entry
2. TTL update entry only -> TTL update entry
3. DELETE entry only -> DELETE entry
4. PUT + DELETE -> DELETE
5. TTL update + DELETE -> DELETE
6. PUT + TTL update -> PUT (the one relevant to this comment)
7. All three -> DELETE

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java [hunk 8]
```diff
@@ -937,6 +940,83 @@ public void orphanedPutRecordsTest() throws Exception {
...
+    assertEquals("All ids not present", ids, seenIds);
+  }
+
+  /**
+   * Tests some recovery scenarios related to TTL update records in particular
+   * @throws Exception
+   */
+  @Test
+  public void ttlUpdateSpecificRecoveryTest() throws Exception {
+    bundleReadBuffer = null;
```
Comment (zzmao) [inline ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: Why `bundleReadBuffer` is null? Looks like they still pass with not null `bundleReadBuffer`.

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 9]
```diff
@@ -590,140 +590,286 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
-            + "copy: {}", indexSegmentStartOffset, storeId, allIndexEntries, copyCandidates, indexEntriesToCopy);
-    return indexEntriesToCopy;
   }
 
   /**
    * Gets all the valid index entries in the given list of index entries.
-   * @param indexSegmentStartOffset the start {@link Offset} of the {@link IndexSegment} that {@code allIndexEntries} is
-   *                                from.
+   * @param indexSegment the {@link IndexSegment} that {@code allIndexEntries} are from.
    * @param allIndexEntries the list of {@link IndexEntry} instances from which the valid entries have to be chosen.
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Done

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java [hunk 10]
```diff
@@ -590,140 +590,286 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
...
    * @return the list of valid entries picked from {@code allIndexEntries}.
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Added comment

File: ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java (hunk 7 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java]: Added a comment saying that this (^) is assumed

File: ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java (hunk 8 above)
Comment (vgkholla) [reply ambry-store/src/test/java/com.github.ambry.store/BlobStoreCompactorTest.java]: Some of the situations induced by the test are more predictable if the `bundleReadBuffer` is not used (the `InterruptionInducing..." impls of various components - esp the `DiskIOScheduler`).
